sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala

```diff
@@ -325,11 +325,17 @@ case class ExpressionEncoder[T](
   assert(serializer.forall(_.references.isEmpty), "serializer cannot reference any attributes.")
   assert(serializer.flatMap { ser =>
     val boundRefs = ser.collect { case b: BoundReference => b }
-    assert(boundRefs.nonEmpty,
-      "each serializer expression should contain at least one `BoundReference`")
+    assert(boundRefs.nonEmpty || isEmptyStruct(ser),
```

Comment:
what if you have empty boundref in nonStruct and in struct as well ?

Reply:
can you maybe elaborate a bit? I'm not 100% sure what you mean

Reply:
The change looks correct to me. If a column is StructType, the only reason that its serialize has no `BoundReference` is it's an empty struct.

Reply:
I'm just saying that `isEmptyyStruct` condition can mask any cases where boundRefs is empty and it shouldn't be if empty struct is present.

I'll let someone with more context decide whether this is ok

Reply:
i could be wrong but it seems to me that the empty struct will always have empty boundref